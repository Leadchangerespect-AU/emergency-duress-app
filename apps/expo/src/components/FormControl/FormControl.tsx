import { cn } from "@/utils/cn";
import React, { createContext, useContext, ReactNode } from "react";
import { View } from "react-native";

interface FormControlContextValue {
  isInvalid?: boolean;
  isDisabled?: boolean;
  isRequired?: boolean;
  //  can add more states like isDisabled, isRequired, etc as needed
}

// Create the context with a default value
const FormControlContext = createContext<FormControlContextValue>({});

// Hook for child components to consume the context value
export const useFormControlContext = () => {
  return useContext(FormControlContext);
};

type FormControlProps = {
  className?: string;
  children: ReactNode;
} & FormControlContextValue;

export const FormControl = ({
  children,
  isInvalid,
  isDisabled,
  isRequired,
  className,
}: FormControlProps) => {
  // Provide state to child components
  return (
    <FormControlContext.Provider value={{ isInvalid, isDisabled, isRequired }}>
      <View className={cn("my-2", className)}>{children}</View>
    </FormControlContext.Provider>
  );
};
