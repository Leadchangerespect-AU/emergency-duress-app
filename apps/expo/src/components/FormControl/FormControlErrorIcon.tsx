import React from "react";
import { Ionicons } from "@expo/vector-icons";
import colors from "tailwindcss/colors";

type IoniconsName = keyof typeof Ionicons.glyphMap;

interface FormControlErrorProps {
  className?: string;

  /**
   * The name of the Ionicons icon.
   * @default "alert-circle-outline"
   */
  name?: IoniconsName;

  /**
   * The size of the icon.
   * @default 24
   */
  size?: number;

  /**
   * The color of the icon.
   * @default colors.red[500]
   */
  color?: string;
}

export const FormControlErrorIcon = ({
  className,
  name = "alert-circle-outline",
  size = 24,
  color = colors.red[500],
}: FormControlErrorProps) => {
  return (
    <Ionicons name={name} size={size} color={color} className={className} />
  );
};
