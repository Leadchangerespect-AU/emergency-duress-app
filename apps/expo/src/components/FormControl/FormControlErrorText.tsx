import { cn } from "@/utils/cn";
import React, { ReactNode } from "react";
import { Text } from "react-native";

interface FormControlErrorProps {
  className?: string;
  children: ReactNode;
}

export const FormControlErrorText = ({
  children,
  className,
}: FormControlErrorProps) => {
  return (
    <Text className={cn("text-red-500 text-sm", className)}>{children}</Text>
  );
};
