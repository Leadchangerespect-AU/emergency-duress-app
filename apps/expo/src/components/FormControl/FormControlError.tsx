import React, { ReactNode } from "react";
import { View } from "react-native";
import { useFormControlContext } from "./FormControl";
import { cn } from "@/utils/cn";

interface FormControlErrorProps {
  className?: string;
  children: ReactNode;
}

export const FormControlError = ({
  children,
  className,
}: FormControlErrorProps) => {
  // Read the current `invalid` state from context
  const { isInvalid, isDisabled } = useFormControlContext();

  // If not invalid, return null to skip rendering
  if (!isInvalid || isDisabled) return null;

  return (
    <View className={cn("flex flex-row items-center mt-2 gap-2", className)}>
      {children}
    </View>
  );
};
