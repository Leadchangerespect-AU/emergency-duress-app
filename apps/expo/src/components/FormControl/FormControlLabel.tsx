import { cn } from "@/utils/cn";
import React, { LabelHTMLAttributes, ReactNode } from "react";
import { Platform, Text, TextProps } from "react-native";
import { useFormControlContext } from "./FormControl";

type WebProps = LabelHTMLAttributes<HTMLLabelElement>;
type NativeProps = TextProps;

type FormControlLabelProps = typeof Platform.OS extends "web"
  ? WebProps
  : NativeProps;

export const FormControlLabel = ({
  children,
  className,
  ...props
}: FormControlLabelProps) => {
  const { isRequired } = useFormControlContext();

  if (Platform.OS === "web") {
    return (
      <label
        className={cn("text-base text-neutral-800 font-semibold", className)}
        {...(props as WebProps)}
      >
        {children}
        {isRequired && <Text className="ml-0.5">*</Text>}
      </label>
    );
  } else {
    return (
      <Text
        className={cn("text-base text-neutral-800 font-semibold", className)}
        {...(props as NativeProps)}
      >
        {children}
        {isRequired && <Text className="ml-0.5">*</Text>}
      </Text>
    );
  }
};
