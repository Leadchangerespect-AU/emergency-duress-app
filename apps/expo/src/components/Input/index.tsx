import React from "react";
import { Platform, TextInput, TextInputProps } from "react-native";
import { tv } from "tailwind-variants";

import { useFormControlContext } from "../FormControl/FormControl";

const input = tv({
  base: "block w-full rounded-md bg-white px-3 py-2.5 text-base text-gray-900 outline outline-1 -outline-offset-1 outline-gray-300 placeholder:text-gray-400 focus:outline focus:outline-2 focus:-outline-offset-2 focus:outline-accent sm:text-sm/6",
  variants: {
    disabled: {
      true: "cursor-not-allowed border-gray-200 text-gray-300 placeholder:text-gray-400",
    },
    invalid: {
      true: "border-red-500",
    },
  },
});

const Input = ({ className, ...props }: TextInputProps) => {
  const { isDisabled, isInvalid, isRequired } = useFormControlContext();
  return (
    <TextInput
      editable={!isDisabled || props.editable}
      selectTextOnFocus={!isDisabled || props.selectTextOnFocus}
      readOnly={isDisabled || props.readOnly}
      {...Platform.select({
        web: {
          disabled: isDisabled,
        },
      })}
      aria-disabled={isDisabled || props["aria-disabled"]}
      aria-invalid={isInvalid}
      aria-required={isRequired}
      className={input({
        disabled: isDisabled,
        invalid: isInvalid && !isDisabled,
        className,
      })}
      {...props}
    />
  );
};

export default Input;
