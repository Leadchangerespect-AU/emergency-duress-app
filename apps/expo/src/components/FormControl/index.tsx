import { FormControl, useFormControlContext } from "./FormControl";
import { FormControlError } from "./FormControlError";
import { FormControlErrorIcon } from "./FormControlErrorIcon";
import { FormControlErrorText } from "./FormControlErrorText";
import { FormControlLabel } from "./FormControlLabel";

export {
  useFormControlContext,
  FormControl,
  FormControlError,
  FormControlErrorIcon,
  FormControlErrorText,
  FormControlLabel,
};
