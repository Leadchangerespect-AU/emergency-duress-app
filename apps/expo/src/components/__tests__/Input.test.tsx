import React from "react";
import { act, render, screen } from "@testing-library/react-native";

import { FormControl } from "../FormControl";
import Input from "../Input";

jest.mock("@expo/vector-icons", () => ({
  Ionicons: "",
}));

describe("Input - Additional Tests", () => {
  it("sets aria-required when isRequired is true", () => {
    render(
      <FormControl isRequired>
        <Input placeholder="Enter text" />
      </FormControl>,
    );

    const input = screen.getByPlaceholderText("Enter text");
    expect(input.props["aria-required"]).toBe(true);
  });

  it("sets aria-disabled when form control is disabled", () => {
    render(
      <FormControl isDisabled>
        <Input placeholder="Enter text" />
      </FormControl>,
    );

    const input = screen.getByPlaceholderText("Enter text");
    expect(input.props["aria-disabled"]).toBe(true);
  });

  it("sets aria-invalid when form control is invalid", () => {
    render(
      <FormControl isInvalid>
        <Input placeholder="Enter text" />
      </FormControl>,
    );

    const input = screen.getByPlaceholderText("Enter text");
    expect(input.props["aria-invalid"]).toBe(true);
  });

  it("does not set aria-invalid when form control is valid", () => {
    render(
      <FormControl>
        <Input placeholder="Enter text" />
      </FormControl>,
    );

    const input = screen.getByPlaceholderText("Enter text");
    expect(input.props["aria-invalid"]).toBeFalsy();
  });

  it("ensures className updates based on validity", () => {
    render(
      <FormControl isInvalid>
        <Input placeholder="Enter text" />
      </FormControl>,
    );

    const input = screen.getByPlaceholderText("Enter text");
    expect(input.props.className).toContain("border-red-500");
  });

  it("ensures className updates based on disabled state", () => {
    render(
      <FormControl isDisabled>
        <Input placeholder="Enter text" />
      </FormControl>,
    );

    const input = screen.getByPlaceholderText("Enter text");
    expect(input.props.className).toContain("border-gray-200");
  });

  it("handles focus and blur events correctly", () => {
    const onFocus = jest.fn();
    const onBlur = jest.fn();

    render(
      <Input placeholder="Enter text" onFocus={onFocus} onBlur={onBlur} />,
    );

    const input = screen.getByPlaceholderText("Enter text");

    // Call onFocus manually
    act(() => {
      input.props.onFocus();
    });
    expect(onFocus).toHaveBeenCalled();

    // Call onBlur manually
    act(() => {
      input.props.onBlur();
    });
    expect(onBlur).toHaveBeenCalled();
  });

  it("matches snapshot with different states", () => {
    const { toJSON } = render(
      <FormControl isInvalid isDisabled>
        <Input placeholder="Enter text" />
      </FormControl>,
    );

    expect(toJSON()).toMatchSnapshot();
  });
});
