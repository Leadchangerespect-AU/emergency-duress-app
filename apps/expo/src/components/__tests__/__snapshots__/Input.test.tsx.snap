// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input - Additional Tests matches snapshot with different states 1`] = `
<View
  className="my-2"
>
  <TextInput
    aria-disabled={true}
    aria-invalid={true}
    className="block w-full rounded-md bg-white px-3 py-2.5 text-base outline outline-1 -outline-offset-1 outline-gray-300 focus:outline focus:outline-2 focus:-outline-offset-2 focus:outline-accent sm:text-sm/6 cursor-not-allowed border-gray-200 text-gray-300 placeholder:text-gray-400"
    placeholder="Enter text"
    readOnly={true}
  />
</View>
`;
